### PR TITLE
fix: Convert DMatrix to CSR matrix if xgboost version allows

### LINKED ISSUE #3357: BUG: TreeExplainer assigns xgboost Dmatrix to Explanation.data
### Issue Description

When `TreeExplainer.__call__()` is passed a DMatrix object (from `xgboost`), it gets stored as such in the resultant `Explanation` object in `Explanation.data`, but our `shap.plots` functions usually assumes `Explanation.data` is an array-like, e.g. it should have the `.shape` attribute available.

See example below.

### Minimal Reproducible Example

```python
import numpy as np
import xgboost
import shap

rs = np.random.RandomState(0)
X = rs.normal(size=(100, 7))
y = np.matmul(X, [-2, 1, 3, 5, 2, 20, -5])

# train a model with single tree
Xd = xgboost.DMatrix(X, label=y)
model = xgboost.train({'eta': 1, 'max_depth': 6, 'base_score': 0, "lambda": 0}, Xd, 1)

explainer = shap.TreeExplainer(model)
explanation = explainer(Xd)

print(explanation.data)  # <xgboost.core.DMatrix at 0x136020940>
shap.plots.beeswarm(explanation)  # AttributeError: 'DMatrix' object has no attribute 'shape'
```

### Traceback

```shell
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[21], line 13
     10 explanation = explainer(Xd)
     12 print(explanation.data)  # <xgboost.core.DMatrix at 0x136020940>
---> 13 shap.plots.beeswarm(explanation)

File ~/Documents/xxx/shap/plots/_beeswarm.py:134, in beeswarm(shap_values, max_display, order, clustering, cluster_threshold, color, axis_color, alpha, show, log_scale, color_bar, plot_size, color_bar_label)
    132         feature_names = features
    133     features = None
--> 134 elif (features is not None) and len(features.shape) == 1 and feature_names is None:
    135     feature_names = features
    136     features = None

AttributeError: 'DMatrix' object has no attribute 'shape'
```

### Expected Behavior

Expected behaviour is that `beeswarm` (or any other plots) should not fail when passed an `Explanation` object.

The cleaner solution should be ensuring the `Explanation.data` doesn't contain an xgboost dmatrix object , rather than modifying all the plots functions to handle dmatrix properly.

### Bug report checklist

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [X] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [X] I'd be interested in making a PR to fix this bug

### Installed Versions

shap master branch c3f2adb439582501c6d0f7ae65a9c56b9d32ce6c

### PR DESCRIPTION
**Pull request by thatlittleboy**
## Overview

Closes #3357  

Description of the changes proposed in this pull request:

- Use the (potentially unstable) `DMatrix.get_data()` method to extract out the data from a DMatrix object into a numpy/scipy array, so that it plays nice with our plotting functions. See linked issue for more background.

## Checklist

- [X] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [X] Unit tests added (if fixing a bug or adding a new feature)

### CODE REVIEW

File: shap/explainers/_tree.py [hunk 1]
```diff
@@ -252,6 +253,21 @@ def __call__(self, X, y=None, interactions=False, check_additivity=True):
...
         if isinstance(X, pd.DataFrame):
             X = X.values
+        elif safe_isinstance(X, "xgboost.core.DMatrix"):
+            import xgboost
+
+            if version.parse(xgboost.__version__) < version.parse("1.7.0"):  # pragma: no cover
+                # cf. GH #3357
+                wmsg = (
+                    "xgboost DMatrix is not supported in `Explanation.data`, setting it to None for "
+                    "now. If you require the `data` attribute (e.g. using `shap.plots`), then either "
```
Comment (connortann) [inline shap/explainers/_tree.py]: I think the first sentence could be rephrased slightly, just to make it clear what is being set to None. How about something like this?

> `shap.Explanation` does not support `xgboost.DMatrix` objects for xgboost < 1.7, so the `data` attribute of the explanation object will be set to `None`. If you require the `data` attribute... 

I think the rest of the message is good.

**Review by connortann — APPROVED** [review]
Approved with an optional minor suggestion. Over to you to merge when happy!

File: shap/explainers/_tree.py (hunk 1 above)
Comment (thatlittleboy) [reply shap/explainers/_tree.py]: Thanks!

**Comment by thatlittleboy** [comment]
@connortann codecov seems to have stopped running, any idea why?

**Comment by connortann** [comment]
Hmm. I noticed this on the CodeCov site, the PR is showing as "Processing":

image

I've found codecov.io to be quite unreliable, and minor outages now and then aren't unusual. I suggest waiting a day or two to see if this problem resolves itself.

**Comment by thatlittleboy** [comment]
@connortann I see, is that view only open for you, or can I also look at that dashboard?

I'll merge this PR in first then, the coverage shouldn't be affected in theory by this PR.

**Comment by connortann** [comment]
That view should be opened to everyone, try here:

<https://app.codecov.io/gh/shap/shap/pulls>

**Pull request merged**